src/bio/bio_recovery.c

```diff
@@ -93,20 +92,35 @@ teardown_blobstore(struct bio_xs_context *xs_ctxt, enum smd_dev_type st)
 	}
 }
 
+struct xs_blobstore_arg {
+	struct bio_xs_context	*xs_ctxt;
+	struct bio_blobstore	*bbs;
+};
+
 static void
 teardown_xstream(void *arg)
 {
-	struct bio_xs_context	*xs_ctxt = arg;
+	struct xs_blobstore_arg *xs_bbs_arg = arg;
+	enum smd_dev_type	 st;
+	struct bio_xs_blobstore *bxb = NULL;
 
-	D_ASSERT(xs_ctxt != NULL);
+	D_ASSERT(xs_bbs_arg->xs_ctxt != NULL);
 	if (!is_server_started()) {
 		D_INFO("Abort xs teardown on server start/shutdown\n");
-		return;
+		goto out;
 	}
 
-	xs_ctxt->bxc_ready = 0;
-	/* only teardown data blobstore for now */
-	teardown_blobstore(xs_ctxt, SMD_DEV_TYPE_DATA);
+	xs_bbs_arg->xs_ctxt->bxc_ready = 0;
+	for (st = SMD_DEV_TYPE_DATA; st < SMD_DEV_TYPE_MAX; st++) {
```

Comment:
hmm, that means if a data device is faulty, we'll teardown wal & meta devices as well? Imagine an engine has 8 targets, 8 data SSDs and 1 wal/meta SSD, if one data SSD is faulty, it should affect only single target (but not all 8 targets).

It's better to have a detailed design for this feature, it looks to me not a trivial work.